### ISSUE TITLE
Docker shouldn't prevent --dns from being used with --net=host

### ISSUE THREAD

**Issue by bboreham** [docker/docker#10384]
Currently, using Docker 1.4.0, I get:

````
$ docker run --net=host --dns=1.2.3.4 ubuntu
FATA[0000] Conflicting options: --net=host can't be used with --dns. This configuration is invalid.
````

I can see that this would be an unusual combination, but I don't see that it is invalid.  It seems I can achieve the effect by creating a file and bind-mounting it - `-v /tmp/myresolv:/etc/resolv.conf`, so don't see why Docker refuses to do this for me.

**Comment by unclejack** [comment]
@bboreham `--net=host` means that the host's networking stack is being used. There were problems and bugs with other setups and this is why it was changed.

**Comment by bboreham** [comment]
Can you remember any more about what happened?  I tried searching the old issues, but didn't come up with anything.

Looks like someone else [had the same thought](https://github.com/docker/docker/commit/e49c70109228d657790190007c8f9d7c55a25be2#commitcomment-7955850), although perhaps attached it to the wrong commit.

**Comment by unclejack** [comment]
@bboreham There were a lot of problems around this. I recall that some users reported some problems with this and it didn't really seem like it was Docker's fault.

**Comment by estesp** [comment]
The original PR that started ignoring DNS settings when `--net=host` is #5129 just for reference.  The warning about conflicting options was added later.  Of course, prior to that PR it would have been ugly to modify resolv.conf in `--net=host` mode as it would have edited the host's resolver config!  After that PR, just like all other containers, each container received a private copy, so now I'm tempted to consider why it makes sense to prevent.  @vieux given you wrote the original PR, any thoughts on this other than a potentially reasonable effort to protect users from doing something silly after they have effectively declared "I want to use the host's network settings/stack as-is"?

**Comment by jamicque** [comment]
There are some applications, like for example Asterisk (open source SIP server) which send their IP addresses in SIP packets depending on IP of the network interface it owns. Sending private IP in SIP packet in SDP part can cause deaf connections. There is a way to overcome this:
- change the configuration of Asterisk saying which IP address should be stated in those packets
However, the problem begins when you have multiple IP addresses depending on the network, than such configuration is not possible.
The way to overcome this obstacle is to state --net="host", but then Asterisk has no access to docker containers by dns name provided by DNS server (skydns for example).
You can change the resolv.conf on your host machine stating the first dns to 127.0.0.1 and then running your docker dns service with exposed ports but it is much more "dirty" than the option of defining the dns server during the start of container with host option.

Asterisk of course is one example, I'm sure there are much more of them...

**Comment by jfrazelle** [comment]
this is working as intended

**Comment by omribahumi** [comment]
@jfrazelle You mean that the intended behavior is to disallow `--dns` with `--net=host` ?
I understand why it was originally introduced (`--net=host` meant the host's `/etc/resolv.conf` was bind-mounted), but this is no longer the case.
I too suffer from this restriction.

Will appreciate your response. Thanks

**Comment by Smana** [comment]
+1

**Comment by kacole2** [comment]
+1 would like to see the ability of using `--add-host` to containers that need to talk on user-defined networks but use the `--net=host` stack

**Comment by Insomnium** [comment]
+1

**Comment by bboreham** [comment]
Fixed in #22408
